File: src/EdgeAIManager.fs
// EdgeAIManager.fs
// Edge-AI-APP
// Created by Alexandre on 11/04/2026
// Example extensive F# file

open System
open System.Text

type EdgeAIManager(modelName : string) =
    let mutable isConfigured = false
    let mutable name = modelName.Trim()

    member this.ModelName
        with get() = name
        and set(value) = name <- value.Trim()

    member this.IsConfigured
        with get() = isConfigured
        and set(value) = isConfigured <- value

    member this.LoadModel() =
        printfn "Loading model: %s" name
        System.Threading.Thread.Sleep(1000)
        printfn "Model %s loaded successfully." name

    member this.Configure(options : Map<string,string>) =
        printfn "Configuring EdgeAIManager with options:"
        options |> Map.iter (fun k v -> printfn "  %s = %s" k v)
        isConfigured <- true

    member this.ProcessInput(input : string) =
        printfn "Processing input of length: %d" input.Length
        let output = new string(Array.rev(input.ToCharArray()))
        printfn "Processing complete."
        output

// -------------------------------
// Usage Example
// -------------------------------
[<EntryPoint>]
let main argv =
    let manager = EdgeAIManager(" edge_ai_model.onnx ")
    manager.LoadModel()
    manager.Configure(Map.ofList [("quantization","int8"); ("device","CPU")])

    let inputData = "HelloEdgeAI"
    let outputData = manager.ProcessInput(inputData)
    printfn "Output result: %s" outputData
    0

File: src/EdgeAIManager.nim
# EdgeAIManager.nim
# Edge-AI-APP
# Created by Alexandre on 11/04/2026
# Example extensive Nim file

import os, strutils, sequtils, times, asyncdispatch, tables

# -------------------------------
# Protocol-like definition (using concepts)
# -------------------------------
type
  EdgeAIProcessing* = concept x
    x.processInput(data: seq[byte]): Future[seq[byte]]
    x.configure(options: Table[string,string])

# -------------------------------
# Extensions (like categories)
# -------------------------------
proc edgeAITrimmed*(s: string): string =
  ## Trim whitespace and newlines
  s.strip(chars = {' ', '\n', '\t'})

# -------------------------------
# Main class equivalent
# -------------------------------
type
  EdgeAIManager* = ref object
    modelName*: string
    isConfigured*: bool

# Constructor
proc newEdgeAIManager*(modelName: string): EdgeAIManager =
  EdgeAIManager(modelName: modelName.edgeAITrimmed, isConfigured: false)

# Load model
proc loadModel*(self: EdgeAIManager) =
  echo "Loading model: ", self.modelName
  sleep(1000) # simulate loading
  echo "Model ", self.modelName, " loaded successfully."

# Configure
proc configure*(self: EdgeAIManager, options: Table[string,string]) =
  echo "Configuring EdgeAIManager with options: ", options
  self.isConfigured = true

# Process input (simulate inference by reversing bytes)
proc processInput*(self: EdgeAIManager, data: seq[byte]): Future[seq[byte]] {.async.} =
  echo "Processing input of length: ", data.len
  var output = data
  for i in 0 ..< output.len div 2:
    let j = output.len - i - 1
    swap(output[i], output[j])
  echo "Processing complete."
  return output

# -------------------------------
# Usage Example
# -------------------------------
when isMainModule:
  let manager = newEdgeAIManager(" edge_ai_model.onnx ")
  manager.loadModel()
  manager.configure({"quantization": "int8", "device": "CPU"}.toTable)

  let inputData = cast[seq[byte]]("HelloEdgeAI")
  let fut = manager.processInput(inputData)
  waitFor fut
  let result = cast[string](fut.read)
  echo "Output result: ", result
